### PR TITLE
Bug1794213 server side keygen enroll

### PR DESCRIPTION
**Pull request by ladycfu**
There are four patches.  It's best to apply them all in the correct order to test out together:

1. (ladycfu) The first patch adds support for  the ServerSideKeygen Enrollment feature ( for RSA only)
2. (SilleBille)The second patch fixes the UI keyType/keySize selection and makes the values go to the backend 
3. (Jmagne)The 3rd patch fixes the default p12 filename, as well as the transport cert nickname 
4. (ladycfu) The 4th patch provides support for ECC, as well as auditing.  It also fixes the "missing request record" issue and some sanitization.

### CODE REVIEW

File: base/ca/shared/profiles/ca/caServerKeygen_UserCert.cfg [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+policyset.userCertSet.2.constraint.params.notBeforeCheck=false
+policyset.userCertSet.2.constraint.params.notAfterCheck=false
+policyset.userCertSet.2.default.class_id=validityDefaultImpl
+policyset.userCertSet.2.default.name=Validity Default
+policyset.userCertSet.2.default.params.range=180
+policyset.userCertSet.2.default.params.startTime=0
+policyset.userCertSet.3.constraint.class_id=keyConstraintImpl
+policyset.userCertSet.3.constraint.name=Key Constraint
+policyset.userCertSet.3.constraint.params.keyType=-
+policyset.userCertSet.3.constraint.params.keyParameters=1024,2048,3072,4096,nistp256,nistp384,nistp521
```
Comment (jmagne) [inline base/ca/shared/profiles/ca/caServerKeygen_UserCert.cfg:40]: I think you need to add my PSS algs in there like I did, just take a look at any of the other latest profile files.

File: base/common/src/com/netscape/certsrv/system/KRAConnectorInfo.java [hunk 2]
```diff
@@ -50,6 +51,9 @@
     @XmlElement
     String transportCert;
 
+    @XmlElement
```
[the remark this answers is not in the document]
Comment (ladycfu) [reply base/common/src/com/netscape/certsrv/system/KRAConnectorInfo.java:54]: That's your own patch.  As I said, there are 4 patches.  One of them is yours.

File: base/kra/src/com/netscape/kra/AsymKeyGenService.java [hunk 3]
```diff
@@ -72,12 +76,48 @@ public AsymKeyGenService(IKeyRecoveryAuthority kra) {
...
+                    keySize = 256;
+                    break;
+                case "nistp384":
+                    keySize = 384;
+                    break;
+                case "nistp521":
+                    keySize = 521;
+                    break;
+                default:
+                    CMS.debug(method + "unknown EC key curve name: " + keySizeStr);
```
[the remark this answers is not in the document]
Comment (ladycfu) [reply base/kra/src/com/netscape/kra/AsymKeyGenService.java:117]: ok

File: base/kra/src/com/netscape/kra/AsymKeyGenService.java [hunk 4]
```diff
@@ -141,22 +181,81 @@ public boolean serviceRequest(IRequest request) throws EBaseException {
...
+            }
+        }
+
+        if (isSSKeygen) {
+            byte[] publicKeyData = null;
+            String pubKeyStr = "";
+            try {
+                publicKeyData = kp.getPublic().getEncoded();
+                if (publicKeyData == null) {
+                    request.setExtData(IRequest.RESULT, Integer.valueOf(4));
```
[the remark this answers is not in the document]
Comment (ladycfu) [reply base/kra/src/com/netscape/kra/AsymKeyGenService.java:231]: It's just an error code.

File: base/kra/src/com/netscape/kra/AsymKeyGenService.java [hunk 5]
```diff
@@ -167,23 +266,46 @@ public boolean serviceRequest(IRequest request) throws EBaseException {
...
         KeyRecord record = new KeyRecord(null, kp.getPublic().getEncoded(), privateSecurityData,
-                owner, algorithm, owner);
+                isSSKeygen? clientKeyId:owner, algorithm, owner);
 
         IKeyRepository storage = kra.getKeyRepository();
         BigInteger serialNo = storage.getNextSerialNumber();
 
         if (serialNo == null) {
-            kra.log(ILogger.LL_FAILURE,
+            if (isSSKeygen) {
```
[the remark this answers is not in the document]
Comment (ladycfu) [reply base/kra/src/com/netscape/kra/AsymKeyGenService.java:293]: this allows audit messages to be enabled/disabled separately.

File: base/server/cms/src/com/netscape/cms/profile/common/CAEnrollProfile.java [hunk 6]
```diff
@@ -209,6 +319,67 @@ public void execute(IRequest request)
 
         request.setExtData(REQUEST_ISSUED_CERT, theCert);
 
+        // cert issued, now retrieve p12
+        if (isSSKeygen) {
```
[the remark this answers is not in the document]
Comment (ladycfu) [reply base/server/cms/src/com/netscape/cms/profile/common/CAEnrollProfile.java:323]: possibly.

File: base/server/cms/src/com/netscape/cms/profile/common/CAEnrollProfile.java [hunk 7]
```diff
@@ -209,6 +319,67 @@ public void execute(IRequest request)
...
+                IConnector kraConnector = caService.getKRAConnector();
+
+                if (kraConnector == null) {
+                    String message = "KRA connector not configured";
+                    CMS.debug(method + message);
+                } else {
+                    CMS.debug(method + "request");
+                    kraConnector.send(request);
+
+                    // check response
```
[the remark this answers is not in the document]
Comment (ladycfu) [reply base/server/cms/src/com/netscape/cms/profile/common/CAEnrollProfile.java:339]: Response is in request.

**Pull request merged**